Add mapAt(keyPath: KeyPath)

michaelavila opened the pull request:
The goal of this PR is to introduce an operator that allows you to map from some value to some other value, given a key path. An example of the usage follows:

```
struct Person {
    let name: String
}

let people: [Person] = [
    Person(name: "Bart"),
    Person(name: "Lisa"),
    Person(name: "Maggie")
]

Observable.from(people)
    .mapAt(\.name)
    .subscribe(onNext: { print($0) })

// Output:
// Bart
// Lisa
// Maggie
```

My organization is probably incorrect. Let me know if there's anything I can change.

freak4pc commented:
I'll try to take a look at this later, but at first look aren't you mapping *from* and not *to* ? `mapFrom(keyPath)` seems more appropriate, or if an overload works here `map(from keyPath: KeyPath)`

michaelavila commented:
@freak4pc maybe, I was thinking of this as mapping to the value at the key path. Are you thinking of it as creating a mapping from a key path? That makes sense. I'm not too concerned about this, so I'll follow your lead if you have a preference.

freak4pc commented:
Oh I meant to open a discussion, not necessarily immediately change :) 

I think the "source" your mapping from is an observable, and the destination is also an observable, nota a keypath. 

So I think my top 3 would be:

- `mapFrom(keyPath: KeyPath)`
- `mapKeyPath(_ keyPath: KeyPath)`
- `map(from keyPath: KeyPath)` - i think this might be bad since we have some bad experiences with overloads of map

@fpillet would love your thoughts

I'll review this code/docs wise a bit later :) 

Thank you @michaelavila !

michaelavila commented:
@freak4pc of those options I like `mapFrom(keyPath: KeyPath)`. Let me know if I should just make that change.

michaelavila commented:
@freak4pc can you elaborate more and why you think `from` makes more sense? Admittedly, I still don't get it.

bobgodwinx commented:
@michaelavila I think @freak4pc was referring to the problem of [map](https://github.com/RxSwiftCommunity/RxSwiftExt/commit/1a7cf0f73dbac48713501ddeee2ccad871ddfa37). IMHO `map(from keyPath: KeyPath)` makes more sense. But if it shows errors use `mapFrom(keyPath: KeyPath)`

fpillet commented:
Since we already have  `mapTo` I think what makes more sense in the context of RxSwiftExt is to use `mapFrom`

freak4pc commented:
keyPath is an index for all sorts
When you map to something - you expect to get what you’re mapping to, but you’re not returning a key path here.

mapAt(keyPath:)

Also makes a lot of sense to me since at is constantly used in Swift for indices @michaelavila

michaelavila commented:
@bobgodwinx , @fpillet, @freak4pc OK. I'm buying this. I'll make the necessary changes and push. Thanks for the feedback.

michaelavila commented:
Done. I went with `mapAt<R>(_ keyPath: KeyPath<E, R>) -> Observable<R>`

freak4pc commented on Playground/RxSwiftExtPlayground.playground/Pages/Index.xcplaygroundpage/Contents.swift:
```diff
@@ -26,6 +26,7 @@
  - [ignore(Any...)](ignore) operator, filters out any of the elements passed in parameters
  - [Observable.once(Element)](once) contructor, creates a sequence that delivers an element *once* to the first subscriber then completes. The same sequence will complete immediately without delivering any element to all further subscribers.
  - [mapTo(Any)](mapTo) operator, takes a sequence of elements and returns a sequence of the same constant provided as a parameter
+ - [mapAt(KeyPath)](mapAt) operator, takes a sequence of elements and returns a sequence of those elements mapped to their values at the given key path
```
Can you move this above mapTo (so its alphabetic within the group of maps)

freak4pc commented on Playground/RxSwiftExtPlayground.playground/Pages/mapAt.xcplaygroundpage/Contents.swift:
```diff
@@ -0,0 +1,43 @@
+
+ The `mapAt` operator takes a sequence of elements and returns a sequence where each element is mapped to its value at the given key path
+ */
+example("replace any input with the value at key path") {
```
replace any input with its value at the provided keypath *

freak4pc commented on Readme.md:
```diff
@@ -181,6 +181,30 @@ next(Nope.)
 
+#### mapAt
+
+Replace every element with its value at the given key path.
```
Its not urgent, so I don't mind merging without change, but I think the term replacing is really confusing - map is a transform method, I would expect something like 

"Transform every element to its value at a given keypath" 

If we change it , then we'll later change this on mapTo as well since I don't think the replace word is appropriate

freak4pc commented on Tests/RxSwift/MapAtTests.swift:
```diff
@@ -0,0 +1,57 @@
+import RxSwiftExt
+import RxTest
+
+struct Person {
```
You can move this struct inside the MapAtTests class - It's not needed outside of the scope of this file/test

freak4pc commented on Tests/RxSwift/MapAtTests.swift:
```diff
@@ -0,0 +1,57 @@
+        scheduler.start()
+    }
+
+    func testReplaceWithResultCount() {
```
What do you think about the naming of the tests? I'm just thinking out loud here if the name is proper to what its testing

testMapAtKeypathCount()
testMapAtKeypathValues() 

Anyways - I don't really mind leaving as is, it doesn't affect anything much, would love your opinion :)

michaelavila commented:
@freak4pc :+1: Let me know what you think. I made the explanations more consistent, changed up some of the wording, and renamed the tests (along with the reording you asked for).

michaelavila commented:
@freak4pc my pleasure. Thanks for the quick responses and feedback. Hopefully I have more to contribute. Take care.
